pkg/engine/configure_task.go

```diff
@@ -0,0 +1,176 @@
+ * limitations under the License.
+ */
+
+package engine
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"os/exec"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	"gopkg.in/yaml.v3"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/NVIDIA/knavigator/pkg/config"
+)
+
+type ConfigureTask struct {
+	BaseTask
+	configureTaskParams
+
+	client *kubernetes.Clientset
+}
+
+type configureTaskParams struct {
+	Nodes   []virtualNode `yaml:"nodes"`
+	Timeout time.Duration `yaml:"timeout"`
+}
+
+type virtualNode struct {
+	Type        string              `yaml:"type" json:"type"`
+	Count       int                 `yaml:"count" json:"count"`
+	Annotations map[string]string   `yaml:"annotations,omitempty" json:"annotations,omitempty"`
+	Labels      map[string]string   `yaml:"labels,omitempty" json:"labels,omitempty"`
+	Conditions  []map[string]string `yaml:"conditions,omitempty" json:"conditions,omitempty"`
+}
+
+func newConfigureTask(log logr.Logger, client *kubernetes.Clientset, cfg *config.Task) (*ConfigureTask, error) {
+	if client == nil {
+		return nil, fmt.Errorf("%s/%s: Kubernetes client is not set", cfg.Type, cfg.ID)
+	}
+
+	task := &ConfigureTask{
+		BaseTask: BaseTask{
+			log:      log,
+			taskType: TaskConfigure,
+			taskID:   cfg.ID,
+		},
+		client: client,
+	}
+
+	if err := task.validate(cfg.Params); err != nil {
+		return nil, err
+	}
+
+	return task, nil
+}
+
+// validate initializes and validates parameters for ConfigureTask
+func (task *ConfigureTask) validate(params map[string]interface{}) error {
+	data, err := yaml.Marshal(params)
+	if err != nil {
+		return fmt.Errorf("failed to parse parameters in %s task %s: %v", task.taskType, task.taskID, err)
+	}
+	if err = yaml.Unmarshal(data, &task.configureTaskParams); err != nil {
+		return fmt.Errorf("failed to parse parameters in %s task %s: %v", task.taskType, task.taskID, err)
+	}
+
+	if task.Timeout == 0 {
+		return fmt.Errorf("%s: missing parameter 'timeout'", task.ID())
+	}
+
+	return nil
+}
+
+// Exec implements Runnable interface
+func (task *ConfigureTask) Exec(ctx context.Context) (err error) {
+	ctx, cancel := context.WithTimeout(ctx, task.Timeout)
+	defer cancel()
+
+	errs := make(chan error)
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	go func() {
+		wg.Wait()
+		close(errs)
+	}()
+
+	go func() {
+		defer wg.Done()
+		if err := task.updateVirtualNodes(ctx); err != nil {
+			errs <- err
+		}
+	}()
+
+	for e := range errs {
+		if e != nil {
+			task.log.Error(e, "configuration error")
+			err = e
+		}
+	}
+
+	return
+}
+
+func (task *ConfigureTask) updateVirtualNodes(ctx context.Context) error {
+	if len(task.Nodes) == 0 {
+		return nil
+	}
+
+	nodeExpr, err := nodes2json(task.Nodes)
+	if err != nil {
+		return err
+	}
+
+	// update helm repo
+	args := []string{"repo", "add", "--force-update", "knavigator", "https://nvidia.github.io/knavigator/helm-charts"}
+
+	task.log.V(4).Info("Updating helm repo")
+
+	if err = runCommand(ctx, task.log, "helm", args); err != nil {
```

Comment:
Would it be possible to use the helm go SDK -- https://helm.sh/docs/topics/advanced/#go-sdk -- instead of relying on the helm CLI?

Reply:
I tried it. It is not reliable.